hapifirmware.ino: Use ArrayLength(pinControl) instead of NUM_DIGITAL+NUM_ANALOG.

james-prior opened the pull request:
Say what you mean, simply and directly.
Let the machine do the dirty work.
[The Elements of Programming Style](https://en.wikipedia.org/wiki/The_Elements_of_Programming_Style#Lessons)

TylerReedMC commented:
James, please stop. You've been making changes for months now that add no functionality and do not advance the project towards it's end goals in any way. Archie was our most productive and knowledgeable embedded systems developer. He got so sick of having his code nitpicked and changed for philosophical reasons that he's left the project. If you have productive code to contribute, you are welcome to do so but please stop changing other peoples code just because it looks better to you or conforms better with a book. Fork the project and fiddle with stuff like that on your own time.

james-prior commented:
@TylerReedMC  wrote:

> James, please stop. 

As you wish.

When we met, I mentioned that our differences were cultural.
The cultural part of open source is difficult for you.
You know that there is something called open source culture,
but it is too different for you to fully grasp let alone accept. 
It is hard to let go of traditional ways. Your culture shock continues.

Your biggest challenge will be to adopt open source culture
for open source projects. I encourage you to learn more about 
open source culture. You know people who are familiar with 
open source culture. Ask them about the culture. Ask them how 
things would be handled in hypothethical (and not so hypothetical) 
situations. Watch their open source projects. 
There are many resources close to you.
The weekly dojos are one such resource.

> You've been making changes for months now that add no functionality ...

Yes! Exactly!

The vast majority of what I did is called [refactoring](https://wikipedia.org/wiki/code_refactoring).
Refactoring does not change the behavior of code.
I was refactoring mostly for readability, baby step by baby step.
Eventually, I would have figured out the higher levels of code,
refactored and rewritten them, and eventually added functionality.

> ... and do not advance the project towards it's end goals in any way. 

Occasionally, as is typical with refactoring, the improvements 
in readability exposed bugs that I fixed and/or asked about. 
Without refactoring, the code will remain hard to understand.
As features are added without refactoring, the code will get harder
and harder to understand, harder to maintain, and harder to add features to.
Also, it will be harder for new people to contribute.

There are more bugs waiting. They are patient.

> Archie was our most productive and knowledgeable embedded systems 
> developer. He got so sick of having his code nitpicked and 
> changed for philosophical reasons that he's left the project. 
> If you have productive code to contribute, you are welcome to do 
> so but please stop changing other peoples code just because it 
> looks better to you or conforms better with a book. 

It is sad that ozarchie quit without a word on github.
We should have had a meaningful discussion.

It is sad that the importance of refactoring is not understood.
Since you do not want refactoring, I suggest that in your 
contribution guidelines, you state that refactoring is not welcome. 
You could even refer to this conversation. 
That would make your wishes clear to potential contributors.

You are fortunate to have Pedro. He has eyes for good code.
He is also learning much and learning fast. 

Good luck.

ozarchie commented:
Hi James,

I am not sure whether you understood or did not see my comments on HAPI.

I did not "quit without a word". I did reply to you directly on the HAPI
engineering thread and discussed these issues, as well as providing steps
to address the development documentation issues.

As a  software developer for over 40 years, I definitely value the input
and results from both open-source and refactoring. I have used and welcomed
both in many projects that are now implemented in millions of realtime
software applications.

I challenge your assertion that Tyler, who has developed and contributed to
open-source projects on source-forge and github since 2013, does not
understand its culture. HAPI, by its nature is open-source, and welcoming
of many software developers from many different backgrounds. Its purpose is
open-source, looking to improve worldwide growing processes and outcomes
and help in a real-world food production problem.

I further challenge that you understand what refactoring means which by
your own quote - "Refactoring does not change the behavior of code.".  Your
reticence to learn what is needed to refactor, including re-compiling and a
justification of your own code changes, shows a naivety in the
understanding of its purpose. Wasting any programmers time on fixing syntax
errors and mis-conceptions arising from a simple recompilation is not part
of the refactoring ethos. Most, if not all, of your changes required
editing before the modified code would either compile or function i.e.
behave the same way as before.

I suggest that you practice before you preach. Contribute to the many
worthwhile projects that are in the public domain rather that attacking the
leaders, such as Tyler, who inspire others to join and contribute.

I WILL NOT comment again on this topic, but will soon return to, and I
hope, further the goals of HAPI. The  ongoing discussion of, and comments
on, programming issues and philosophy is better suited to conferences and
meetings where many diverse opinions can be canvassed and opinions
coalesced.

Archie

On Fri, Jun 23, 2017 at 10:10 PM, James Prior <notifications@github.com>
wrote:

> @TylerReedMC <https://github.com/tylerreedmc> wrote:
>
> James, please stop.
>
> As you wish.
>
> When we met, I mentioned that our differences were cultural.
> The cultural part of open source is difficult for you.
> You know that there is something called open source culture,
> but it is too different for you to fully grasp let alone accept.
> It is hard to let go of traditional ways. Your culture shock continues.
>
> Your biggest challenge will be to adopt open source culture
> for open source projects. I encourage you to learn more about
> open source culture. You know people who are familiar with
> open source culture. Ask them about the culture. Ask them how
> things would be handled in hypothethical (and not so hypothetical)
> situations. Watch their open source projects.
> There are many resources close to you.
> The weekly dojos are one such resource.
>
> You've been making changes for months now that add no functionality ...
>
> Yes! Exactly!
>
> The vast majority of what I did is called refactoring
> <https://wikipedia.org/wiki/code_refactoring>.
> Refactoring does not change the behavior of code.
> I was refactoring mostly for readability, baby step by baby step.
> Eventually, I would have figured out the higher levels of code,
> refactored and rewritten them, and eventually added functionality.
>
> ... and do not advance the project towards it's end goals in any way.
>
> Occasionally, as is typical with refactoring, the improvements
> in readability exposed bugs that I fixed and/or asked about.
> Without refactoring, the code will remain hard to understand.
> As features are added without refactoring, the code will get harder
> and harder to understand, harder to maintain, and harder to add features
> to.
> Also, it will be harder for new people to contribute.
>
> There are more bugs waiting. They are patient.
>
> Archie was our most productive and knowledgeable embedded systems
> developer. He got so sick of having his code nitpicked and
> changed for philosophical reasons that he's left the project.
> If you have productive code to contribute, you are welcome to do
> so but please stop changing other peoples code just because it
> looks better to you or conforms better with a book.
>
> It is sad that ozarchie quit without a word on github.
> We should have had a meaningful discussion.
>
> It is sad that the importance of refactoring is not understood.
> Since you do not want refactoring, I suggest that in your
> contribution guidelines, you state that refactoring is not welcome.
> You could even refer to this conversation.
> That would make your wishes clear to potential contributors.
>
> You are fortunate to have Pedro. He has eyes for good code.
> He is also learning much and learning fast.
>
> Good luck.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mayaculpa/hapi/pull/115#issuecomment-310649258>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AGLdJ5pPpE_J5qj3J2Nv7x0sLsGEH5SGks5sG6s9gaJpZM4OAxmS>
> .
>
